### ISSUE TITLE
Usage with redis-sentinel

### ISSUE THREAD

**Issue by adame21** [OptimalBits/bull#1306]
So I have a working queue using Bull and it's great.

I have been instructed by my organization that we MUST use redis-sentinel.

Since I'm new to redis and queues, I am asking here just in case this is a painless fix:

What is the simplest way to 'switch' from normal bull-redis to bull-redis-sentinel?
Can it be done without re-factoring everything?
Thanks in advance. amazing package.

**Comment by manast** [comment]
In theory it should work out of the box by using the proper ioredis options, however, cluster mode is known to not work, on the other hand it does not give so much since a queue needs to have all his data in the same node in order to work properly. https://github.com/luin/ioredis#sentinel
Please let us know how it goes.

**Comment by adame21** [comment]
Thanks a lot for answering again.
I believe I managed to get it right but an unrelated issue is preventing me from testing.
I will be back to work tomorrow and hopefully fix the issue and find out if I got it right.
Please leave open and I will update tomorrow, possibly making a small "guide" for future questions about this - of course in case I actually got it right.
Thanks!

**Comment by adame21** [comment]
Update:
I have been trying to make it work using createClient parameter, but it does not process jobs that way.
I'm probably missing something about ioredis.
I really wanted to make it work and make a guide for it but eventually my organization decided to give up on sentinels for now.
Thanks for waiting on me, hopefully in the future I will be able to make it work and write a guide.
Closing for now

**Comment by PendicGordan** [comment]
Was someone able to solve the problem? Are there some workarounds? Any info?

**Comment by elucidsoft** [comment]
So I have spent a considerable amount of time with this and found some odd things that is very confusing to me.  

I was using queue.getJotCounts() as a health check, and it was returning healthy but when I was calling queue.add it would throw an exception: "IORedis: All sentinels are unreachable. Will try again from scratch in 200ms" which is an error that ioredis is throwing.  I found this odd because if I ran queue.getJobCounts again I was getting a proper response, with data from redis so it was obviously able to connect.  It's boggling my mind as to what's going on here, something very odd is occurring.  I think it might be a specific implementation detail for Bull at this point as everyone seems to be using Sentinel just fine with ioredis.  

I may abandon Sentinel because of this issue, which is unfortunate cause it took me a bit to get a stable Sentinel going in a Kubernetes cluster.  I may try to use twemproxy as I've heard many people prefer it over Sentinel, I could also just use both as well apparently so I might keep sentinel AND use twemproxy.  I'm gonna dig around in the code and see if I can spot why its behaving differently...

**Comment by adame21** [comment]
Hey, I'm back here after almost a year, and now we're trying to use Redis sentinel again.
@elucidsoft Thanks for the robust explanation, I am getting the exact same issue right now. 
Did you manage to get anywhere with this issue?
Thanks

**Comment by adame21** [comment]
Update:
My issue was related to the organization.
I was basically given a non-sentinel Redis server and told it was a sentinel server.
After propely figuring out the connection info I needed, it works like a charm.
I'm giving a code snippet to show how I got it working: 
````
const Queue = require('bull')
const Redis = require("ioredis");

const redisConfig = {
    sentinels: [{
        host: "redacted",
        port: 26379
    }],
    name: "redacted"
}

var myCreateClient = () => {
    return new Redis(redisConfig)
}

const queue = new Queue("clusterTest2", {
    createClient: myCreateClient
})


queue.process((job, done) => {
    console.log("job happening");

    try {
        console.log(`job #${job.id}`);
        console.log(`${JSON.stringify(job.data)}`)
        done()
    }
    catch (err) {
        console.log(`Failed job #${job.id}`);
        done(err)
    }
})
````
Hopefuly someone finds this useful someday :)

**Comment by Able1991** [comment]
> Update:
> My issue was related to the organization.
> I was basically given a non-sentinel Redis server and told it was a sentinel server.
> After propely figuring out the connection info I needed, it works like a charm.
> I'm giving a code snippet to show how I got it working:
> 
> ```
> const Queue = require('bull')
> const Redis = require("ioredis");
> 
> const redisConfig = {
>     sentinels: [{
>         host: "redacted",
>         port: 26379
>     }],
>     name: "redacted"
> }
> 
> var myCreateClient = () => {
>     return new Redis(redisConfig)
> }
> 
> const queue = new Queue("clusterTest2", {
>     createClient: myCreateClient
> })
> 
> 
> queue.process((job, done) => {
>     console.log("job happening");
> 
>     try {
>         console.log(`job #${job.id}`);
>         console.log(`${JSON.stringify(job.data)}`)
>         done()
>     }
>     catch (err) {
>         console.log(`Failed job #${job.id}`);
>         done(err)
>     }
> })
> ```
> 
> Hopefuly someone finds this useful someday :)

Hi. You have no problem with the number of open connections? does everything work well when changing master?

**Comment by adame21** [comment]
Hi, I did run into a silly issue simply because I did not close connections properly.
I opened another issue and was directed properly here:
https://github.com/OptimalBits/bull/issues/1859
Other than that we've been running with redis-sentinel for months without problems :)
Good luck.

**Comment by ericrv01** [comment]
Hello, could you please document how you close the redis connection, I'm not able to figure out it can be done ?
Thanks
